tools/hlsl_generator/gen.py

```diff
@@ -0,0 +1,311 @@
+import json
+import io
+import os
+import re
+from enum import Enum
+from argparse import ArgumentParser
+from typing import NamedTuple
+from typing import Optional
+
+head = """// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_SPIRV_INTRINSICS_CORE_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SPIRV_INTRINSICS_CORE_INCLUDED_
+
+#ifdef __HLSL_VERSION
+#include "spirv/unified1/spirv.hpp"
+#include "spirv/unified1/GLSL.std.450.h"
+#endif
+
+#include "nbl/builtin/hlsl/type_traits.hlsl"
+
+namespace nbl 
+{
+namespace hlsl
+{
+#ifdef __HLSL_VERSION
+namespace spirv
+{
+
+//! General Decls
+template<uint32_t StorageClass, typename T>
+using pointer_t = vk::SpirvOpaqueType<spv::OpTypePointer, vk::Literal< vk::integral_constant<uint32_t, StorageClass> >, T>;
```

Comment:
`pointer_t` should be done in terms of 
```cpp
template<uint32_t StorageClass, typename T>
struct pointer
{
   using type = vk::SpirvOpaqueType<spv::OpTypePointer, vk::Literal< vk::integral_constant<uint32_t, StorageClass> >, T>;
};
// partial spec for BDA
template<typename T>
struct pointer<StorageClassPhysicalStorageBuffer,T>
{
   using type = vk::SpirvType<spv::OpTypePointer,sizeof(uint64_t),sizeof(uint64_t),vk::Literal< vk::integral_constant<uint32_t, StorageClass> >, T>;
};

template<uint32_t StorageClass, typename T>
using pointer_t = pointer::type;
```

and also have a `is_pointer::value` tester + `is_pointer_v`

Reply:
this will go into `type_constraints.hlsl`. right?

Reply:
no it will go here because its `spirv` specific, it can be built with type_traits though, such as `is_same_v`